На ветке lesson9

RamzanBersanov opened the pull request:
Изменения, которые будут включены в коммит: задание 9

RamzanBersanov committed:
На ветке lesson9
 Изменения, которые будут включены в коммит: задание 9

RamzanBersanov committed:
На ветке lesson9
 Изменения, которые будут включены в коммит:
	изменено:      app/models/test_passage.rb

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -0,0 +1,29 @@
+
+  def show
+  
+  end
```
def show; end

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -0,0 +1,29 @@
+
+  def result
+
+  end
```
def result; end

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -1,20 +1,22 @@
 class TestsController < ApplicationController
+
+  before_action :set_user, only: :start
```
для 1 экшена колбек не объявляют. Это усложняет код и дебажить будет сложнее

VidgarVii commented on app/models/test_passage.rb:
```diff
@@ -0,0 +1,55 @@
+  belongs_to :current_question, class_name: 'Question', foreign_key: 'current_question_id', optional: true
+
+  before_validation :before_validation_set_first_question, on: :create
+  before_update :before_update_set_next_question
```
эти два коллбека делают одно и то же по сути
оба устанавливают следующий вопрос
разница лишь в том, что считать следующим вопросом, т.е разница в поиске этого следующего вопроса

(и обрати внимание - оба вызываются с дополнительным условием, на создании или на апдейте)

имеет смысл объединить их в один простой коллбек:

```ruby
before_validation :set_current_question

...

def set_current_question
  self.current_question = next_question
end
```
а всю различающуюся логику поместить уже в метод `next_question`

(коллбеки бывает удобно использовать, но они часто усложняют код и его отладку, если есть возможность не плодить лишние, то ей надо воспользоваться)


вообще с коллбеками нужно осторожно - они не всегда делают то что ты ожидаешь, тогда когда ты ожидаешь ))
бывают ньюансы о которых не догадываешься, особенно при совместной работе нескольких коллбеков и использовании коллбеков с условиями
поэтому если есть возможность упростить коллбеки, то это нужно делать

VidgarVii commented on app/models/test_passage.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  def correct_answer?(answer_ids)
+    correct_answers.ids.sort == answer_ids.map(&:to_i).sort
```
здесь хорошо бы учесть ситуацию когда в `answer_ids` прийдет `nil`
код при этом не должен падать

VidgarVii commented on app/views/test_passages/result.html.erb:
```diff
@@ -0,0 +1,11 @@
+
+<p><%= "Количество правильных ответов: #{@test_passage.correct_questions}" %></p>
+<p><%= "Дано #{@test_passage.correct_percentage}% правильных ответов" %></p>
+<% if @test_passage.correct_percentage >= 85 %>
```
вот эта операция сравнения - это по сути бизнес-логика о которой вьюхе знать не надо и ее правильно будет поместить в модель, а здесь оставить что-то типа:

```html
  <% if @test_passage.successful? %>
```

VidgarVii replied:
лучше "магические цифры" хранить в константах с "говорящими именами"
например `SUCCESS_RATIO`

VidgarVii commented on app/views/test_passages/result.html.erb:
```diff
@@ -0,0 +1,11 @@
+<p><%= "Количество правильных ответов: #{@test_passage.correct_questions}" %></p>
+<p><%= "Дано #{@test_passage.correct_percentage}% правильных ответов" %></p>
+<% if @test_passage.correct_percentage >= 85 %>
+<p style="color: green;">Тест пройден успешно!</p>
```
стилизацию лучше делать через `css` классы:
создать классы, к примеру `test_passage.succes`, `test_passage.fail` и там все описать

VidgarVii commented on config/routes.rb:
```diff
@@ -1,5 +1,18 @@
+
+    member do
+      post :start
+    end
```
когда всего одна строчка, можно написать
```ruby
post :start, on: :member
```

RamzanBersanov committed:
На ветке lesson9
 Изменения, которые будут включены в коммит:
	новый файл:    app/assets/stylesheets/test_passages.scss
	изменено:      app/controllers/test_passages_controller.rb
	изменено:      app/controllers/tests_controller.rb
	изменено:      app/models/test_passage.rb
	изменено:      app/views/test_passages/result.html.erb
	изменено:      config/routes.rb

VidgarVii commented on app/models/test_passage.rb:
```diff
@@ -0,0 +1,55 @@
+
+  before_validation :set_current_question, on: [ :create, :update ]
+
+  SUCCESS_RATIO = 85
```
константы пишут в самом вверху класса

RamzanBersanov committed:
На ветке lesson9
 Изменения, которые будут включены в коммит:
	изменено:      app/models/test_passage.rb
